File: reference-python/freshness.py
"""Deterministic cache-freshness logic for confirmed SUTS jurisdiction codes.

Pure date math, no I/O. The skill caches a confirmed code with the date it was
confirmed; this module decides whether that cache entry is still trustworthy.

A cached code is fresh as of `as_of` iff BOTH:
  1. it is within the 90-day TTL: (as_of - confirmed).days <= TTL_DAYS, and
  2. it was confirmed on or after the most recent rate-reset boundary that is
     on or before `as_of` (Colorado local rates change Jan 1 / Jul 1).

`expires_on(confirmed)` is the earlier of confirmed+TTL and the next reset
boundary strictly after confirmed.
"""

from __future__ import annotations

from datetime import date, timedelta

TTL_DAYS = 90

# Colorado local sales-tax rate changes take effect Jan 1 and Jul 1 by statute.
# (month, day) pairs — change these if the statutory boundaries ever move.
RESET_BOUNDARIES = ((1, 1), (7, 1))


def _boundaries_near(year: int) -> list[date]:
    """All reset boundary dates for the year before/of/after `year`, sorted."""
    return sorted(
        date(y, m, d)
        for y in (year - 1, year, year + 1)
        for (m, d) in RESET_BOUNDARIES
    )


def last_reset(d: date) -> date:
    """Most recent reset boundary on or before `d`."""
    return max(b for b in _boundaries_near(d.year) if b <= d)


def next_reset(d: date) -> date:
    """Earliest reset boundary strictly after `d`."""
    return min(b for b in _boundaries_near(d.year) if b > d)


def is_fresh(confirmed: date, as_of: date) -> bool:
    """True iff a code confirmed on `confirmed` is still fresh as of `as_of`."""
    return (as_of - confirmed).days <= TTL_DAYS and confirmed >= last_reset(as_of)


def expires_on(confirmed: date) -> date:
    """The date a code confirmed on `confirmed` stops being fresh: the earlier
    of confirmed+TTL and the next reset boundary strictly after confirmed."""
    return min(confirmed + timedelta(days=TTL_DAYS), next_reset(confirmed))


def check_freshness(confirmed_date: str, as_of: str | None = None) -> dict:
    """String-in/string-out wrapper for the MCP tool. Dates are ISO (YYYY-MM-DD);
    `as_of` defaults to today. Returns {fresh, reason, expires_on}."""
    confirmed = date.fromisoformat(confirmed_date)
    ref = date.fromisoformat(as_of) if as_of else date.today()

    age = (ref - confirmed).days
    reset = last_reset(ref)
    exp = expires_on(confirmed)

    if age > TTL_DAYS:
        fresh, reason = False, (
            f"stale - confirmed {confirmed} is {age} days old, "
            f"exceeds the {TTL_DAYS}-day TTL"
        )
    elif confirmed < reset:
        fresh, reason = False, (
            f"stale - a rate reset on {reset} occurred after the code was "
            f"confirmed ({confirmed})"
        )
    else:
        fresh, reason = True, (
            f"fresh - confirmed {confirmed} is {age} days old, within the "
            f"{TTL_DAYS}-day TTL and on/after the {reset} reset"
        )

    return {"fresh": fresh, "reason": reason, "expires_on": exp.isoformat()}

File: src/freshness.ts
/**
 * Deterministic cache-freshness logic for confirmed SUTS jurisdiction codes.
 * Port of the Python `freshness.py` — pure date math, no I/O. See CONTRACT.md.
 *
 * Fresh as of `asOf` iff BOTH:
 *   1. (asOf - confirmed).days <= TTL_DAYS, and
 *   2. confirmed >= lastReset(asOf)  (Colorado rates reset Jan 1 / Jul 1).
 */

export const TTL_DAYS = 90;

// Colorado local rate changes take effect Jan 1 and Jul 1 by statute.
// [month, day] — change here if the statutory boundaries ever move.
export const RESET_BOUNDARIES: [number, number][] = [
  [1, 1],
  [7, 1],
];

const DAY_MS = 86_400_000;

function parseISO(s: string): Date {
  const [y, m, d] = s.split("-").map(Number);
  if (!y || !m || !d) throw new Error(`invalid date: ${s}`);
  return new Date(Date.UTC(y, m - 1, d));
}

function toISO(d: Date): string {
  return d.toISOString().slice(0, 10);
}

function addDays(d: Date, n: number): Date {
  return new Date(d.getTime() + n * DAY_MS);
}

function daysBetween(a: Date, b: Date): number {
  return Math.round((a.getTime() - b.getTime()) / DAY_MS);
}

function boundariesNear(year: number): Date[] {
  const out: Date[] = [];
  for (const y of [year - 1, year, year + 1]) {
    for (const [m, d] of RESET_BOUNDARIES) out.push(new Date(Date.UTC(y, m - 1, d)));
  }
  return out.sort((a, b) => a.getTime() - b.getTime());
}

export function lastReset(d: Date): Date {
  const before = boundariesNear(d.getUTCFullYear()).filter((b) => b.getTime() <= d.getTime());
  return before[before.length - 1];
}

export function nextReset(d: Date): Date {
  return boundariesNear(d.getUTCFullYear()).find((b) => b.getTime() > d.getTime())!;
}

export function expiresOn(confirmed: Date): Date {
  const ttl = addDays(confirmed, TTL_DAYS);
  const nr = nextReset(confirmed);
  return ttl.getTime() <= nr.getTime() ? ttl : nr;
}

export interface FreshnessResult {
  fresh: boolean;
  reason: string;
  expires_on: string;
}

export function checkFreshness(confirmedDate: string, asOf?: string): FreshnessResult {
  const confirmed = parseISO(confirmedDate);
  const ref = asOf ? parseISO(asOf) : parseISO(toISO(new Date()));

  const age = daysBetween(ref, confirmed);
  const reset = lastReset(ref);
  const exp = expiresOn(confirmed);

  let fresh: boolean;
  let reason: string;
  if (age > TTL_DAYS) {
    fresh = false;
    reason = `stale - confirmed ${confirmedDate} is ${age} days old, exceeds the ${TTL_DAYS}-day TTL`;
  } else if (confirmed.getTime() < reset.getTime()) {
    fresh = false;
    reason = `stale - a rate reset on ${toISO(reset)} occurred after the code was confirmed (${confirmedDate})`;
  } else {
    fresh = true;
    reason = `fresh - confirmed ${confirmedDate} is ${age} days old, within the ${TTL_DAYS}-day TTL and on/after the ${toISO(reset)} reset`;
  }
  return { fresh, reason, expires_on: toISO(exp) };
}
